agenta-backend/agenta_backend/middlewares/api_key_manager.py

```diff
@@ -0,0 +1,229 @@
+import os
+import redis
+import logging
+from datetime import datetime
+from cryptography.fernet import Fernet
+from fastapi import Request, HTTPException
+from agenta_backend.config import settings
+from agenta_backend.models.db_engine import DBEngine
+from agenta_backend.models.db_models import APIKeyDB
+
+engine = DBEngine().engine()
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+
+encryption_key = os.environ["ENCRYPTION_KEY"]
+fernet = Fernet(encryption_key)
+
+
+
+class APIKeyManager:
+    """
+    A class that manages API keys.
+
+    Methods:
+    - create_api_key(user_id, rate_limit, expiration_date=None): Creates a new API key with an encrypted user ID, rate limit, and optional expiration date.
+    - update_api_key(key, api_key): Updates an existing API key with the provided key and APIKeyDB instance.
+    - delete_api_key(key): Deletes an API key with the provided key.
+    - is_valid_api_key(key): Checks if an API key is valid by verifying that it is not blacklisted and not expired.
+    - use_api_key(key): Uses an API key and checks rate limiting.
+    """
+
+    def __init__(self):
+        self.redis_client = redis.from_url(url=settings.redis_url)  
+
+    def create_api_key(self, user_id, rate_limit, expiration_date=None):
+        """
+        Creates a new API key with an encrypted user ID, rate limit, and optional expiration date.
+
+        Args:
+        - user_id: The user ID to be encrypted and used as the API key.
+        - rate_limit: The rate limit for the API key.
+        - expiration_date: Optional expiration date for the API key.
+
+        Raises:
+        - ValueError: If a unique key cannot be generated after 5 trials.
+        """
+        for _ in range(5):
+            # Encrypt the user_id using Fernet encryption
+            encrypted_user_id = fernet.encrypt(user_id.encode())
```

Comment:
We don't need to use fernet. We need to use a hashing function instead (an encryption function that you can't decrypt)